File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/AccountsResource.java
package work.socialhub.kmastodon.api;

import mastodon4j.Range;
import mastodon4j.entity.Account;
import mastodon4j.entity.Relationship;
import mastodon4j.entity.Status;
import mastodon4j.entity.share.Response;

/**
 * @author hecateball
 */
public interface AccountsResource {

    /**
     * Getting the current user.
     *
     * @return the authenticated user's Account
     */
    public Response<Account> verifyCredentials();

    /**
     * Updating the current user.
     *
     * @param displayName the name to display in the user's profile
     * @param note        a new biography for the user
     * @param avatar      a base64 encoded image to display as the user's avatar (e.g.
     *                    data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUoAAADrCAYAAAA...)
     * @param header      a base64 encoded image to display as the user's header image (e.g.
     *                    data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUoAAADrCAYAAAA...)
     * @return the updated Account
     */
    public Response<Account> updateCredentials(String displayName, String note, String avatar, String header);

    /**
     * Fetching an account.
     *
     * @param id
     * @return an Account
     */
    public Response<Account> getAccount(String id);

    /**
     * Getting an account's followers.
     *
     * @param id
     * @return an array of Accounts
     */
    public Response<Account[]> getFollowers(String id);

    /**
     * Getting an account's followers.
     *
     * @param id
     * @param range
     * @return an array of Accounts
     */
    public Response<Account[]> getFollowers(String id, Range range);

    /**
     * Getting who account is following.
     *
     * @param id
     * @return an array of Accounts
     */
    public Response<Account[]> getFollowing(String id);

    /**
     * Getting who account is following.
     *
     * @param id
     * @param range
     * @return an array of Accounts
     */
    public Response<Account[]> getFollowing(String id, Range range);

    /**
     * Getting an account's statuses.
     *
     * @param id
     * @return an array of Statuses
     */
    public Response<Status[]> getStatuses(String id);

    /**
     * Getting an account's statuses.
     *
     * @param id
     * @param range
     * @return an array of Statuses
     */
    public Response<Status[]> getStatuses(String id, Range range);

    /**
     * Getting an account's statuses.
     *
     * @return an array of Statuses
     */
    public Response<Status[]> getStatuses(
            String id,
            boolean onlyPinned,
            boolean onlyMedia,
            boolean excluedeReplies,
            boolean excludeReblogs,
            Range range);

    /**
     * Following an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> follow(String id);

    /**
     * Unfollowing an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> unfollow(String id);

    /**
     * Blocking an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> block(String id);

    /**
     * Unblocking an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> unblock(String id);

    /**
     * Muting an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> mute(String id);

    /**
     * Unmuting an account.
     *
     * @param id
     * @return the target account's Relationship.
     */
    public Response<Relationship> unmute(String id);

    /**
     * Getting an account's relationships.
     *
     * @param id
     * @param ids
     * @return an array of Relationships of the current user to a list of given accounts.
     */
    public Response<Relationship[]> relationships(String id, String... ids);

    /**
     * Searching for accounts.
     *
     * @param query what to search for
     * @return an array of matching Accounts. Will lookup an account remotely if the search term is in the
     * username@domain format and not yet in the database.
     */
    public Response<Account[]> search(String query);

    /**
     * Searching for accounts.
     *
     * @param query what to search for
     * @param limit maximum number of matching accounts to return (default: 40)
     * @return an array of matching Accounts. Will lookup an account remotely if the search term is in the
     * username@domain format and not yet in the database.
     */
    public Response<Account[]> search(String query, long limit);
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/AccountsResource.kt
package work.socialhub.kmastodon.api

import work.socialhub.kmastodon.api.request.accounts.AccountsAccountRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsBlockRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsFollowRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsFollowersRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsFollowingRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsMuteRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsRelationshipsRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsSearchRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsStatusesRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsUnblockRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsUnfollowRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsUnmuteRequest
import work.socialhub.kmastodon.api.request.accounts.AccountsUpdateCredentialsRequest
import work.socialhub.kmastodon.api.response.Response
import work.socialhub.kmastodon.api.response.accounts.AccountsAccountResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsBlockResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsFollowResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsFollowersResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsFollowingResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsMuteResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsRelationshipsResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsSearchResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsStatusesResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsUnblockResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsUnfollowResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsUnmuteResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsUpdateCredentialsResponse
import work.socialhub.kmastodon.api.response.accounts.AccountsVerifyCredentialsResponse
import kotlin.js.JsExport

@JsExport
interface AccountsResource {

    /**
     * Getting the current user.
     */
    suspend fun verifyCredentials(
    ): Response<AccountsVerifyCredentialsResponse>

    @JsExport.Ignore
    fun verifyCredentialsBlocking(
    ): Response<AccountsVerifyCredentialsResponse>

    /**
     * Updating the current user.
     */
    suspend fun updateCredentials(
        request: AccountsUpdateCredentialsRequest
    ): Response<AccountsUpdateCredentialsResponse>

    @JsExport.Ignore
    fun updateCredentialsBlocking(
        request: AccountsUpdateCredentialsRequest
    ): Response<AccountsUpdateCredentialsResponse>

    /**
     * Fetching an account.
     */
    suspend fun account(
        request: AccountsAccountRequest
    ): Response<AccountsAccountResponse>

    @JsExport.Ignore
    fun accountBlocking(
        request: AccountsAccountRequest
    ): Response<AccountsAccountResponse>

    /**
     * Getting an account's followers.
     */
    suspend fun followers(
        request: AccountsFollowersRequest
    ): Response<Array<AccountsFollowersResponse>>

    @JsExport.Ignore
    fun followersBlocking(
        request: AccountsFollowersRequest
    ): Response<Array<AccountsFollowersResponse>>

    /**
     * Getting who account is following.
     */
    suspend fun following(
        request: AccountsFollowingRequest
    ): Response<Array<AccountsFollowingResponse>>

    @JsExport.Ignore
    fun followingBlocking(
        request: AccountsFollowingRequest
    ): Response<Array<AccountsFollowingResponse>>

    /**
     * Getting an account's statuses.
     */
    suspend fun statuses(
        request: AccountsStatusesRequest
    ): Response<Array<AccountsStatusesResponse>>

    @JsExport.Ignore
    fun statusesBlocking(
        request: AccountsStatusesRequest
    ): Response<Array<AccountsStatusesResponse>>

    /**
     * Following an account.
     */
    suspend fun follow(
        request: AccountsFollowRequest
    ): Response<AccountsFollowResponse>

    @JsExport.Ignore
    fun followBlocking(
        request: AccountsFollowRequest
    ): Response<AccountsFollowResponse>

    /**
     * Unfollowing an account.
     */
    suspend fun unfollow(
        request: AccountsUnfollowRequest
    ): Response<AccountsUnfollowResponse>

    @JsExport.Ignore
    fun unfollowBlocking(
        request: AccountsUnfollowRequest
    ): Response<AccountsUnfollowResponse>

    /**
     * Blocking an account.
     */
    suspend fun block(
        request: AccountsBlockRequest
    ): Response<AccountsBlockResponse>

    @JsExport.Ignore
    fun blockBlocking(
        request: AccountsBlockRequest
    ): Response<AccountsBlockResponse>

    /**
     * Unblocking an account.
     */
    suspend fun unblock(
        request: AccountsUnblockRequest
    ): Response<AccountsUnblockResponse>

    @JsExport.Ignore
    fun unblockBlocking(
        request: AccountsUnblockRequest
    ): Response<AccountsUnblockResponse>

    /**
     * Muting an account.
     */
    suspend fun mute(
        request: AccountsMuteRequest
    ): Response<AccountsMuteResponse>

    @JsExport.Ignore
    fun muteBlocking(
        request: AccountsMuteRequest
    ): Response<AccountsMuteResponse>

    /**
     * Unmuting an account.
     */
    suspend fun unmute(
        request: AccountsUnmuteRequest
    ): Response<AccountsUnmuteResponse>

    @JsExport.Ignore
    fun unmuteBlocking(
        request: AccountsUnmuteRequest
    ): Response<AccountsUnmuteResponse>

    /**
     * Getting an account's relationships.
     */
    suspend fun relationships(
        request: AccountsRelationshipsRequest
    ): Response<Array<AccountsRelationshipsResponse>>

    @JsExport.Ignore
    fun relationshipsBlocking(
        request: AccountsRelationshipsRequest
    ): Response<Array<AccountsRelationshipsResponse>>

    /**
     * Searching for accounts.
     */
    suspend fun search(
        request: AccountsSearchRequest
    ): Response<Array<AccountsSearchResponse>>

    @JsExport.Ignore
    fun searchBlocking(
        request: AccountsSearchRequest
    ): Response<Array<AccountsSearchResponse>>
}
